README.md
# connFu Node.js DSL [![Build Status](https://secure.travis-ci.org/bluevialabs/connfu-node.png)](http://travis-ci.org/bluevialabs/connfu-node)

connfu package provides an easy way to get access to [connFu platform](http://connfu.com) using Node.js

## How to install

The eassiest way is by installing it from the `npm` repository:

    
    $ npm install connfu

If you'd prefer to install the latest master version of connFu, you can clone the `connfu-node` source repository
from GitHub and then install it using `npm`:

    
    $ git clone "https://github.com/bluevialabs/connfu-node.git"
    
    $ npm install -g connfu-node/

## How to use it

### Connecting to streams

First, require `connfu` package:

    
    var connFu = require('connfu');

Next, create a basic application using your connFu API KEY:

    
    var app = connFu.createApp('<your_api_key>');

Then, attach some event handlers:

    
    app.on("voice", {
      join: function(params){
        console.log("The number " + params.from + " joined to the conference with number " + params.to);
      },
      leave: function(params){
        console.log("The number " + params.from + " left the conference with number " + params.to);
      }
    });

Finally, just listen!

    
    app.listen();

### Using the CLI

When you install connFu node package, you get a CLI that allows you to do some common tasks fast and easily:

![CLI interface](http://bluevialabs.github.com/connfu-node/images/screenshot_cli.png)

### Sending commands

The `connFu` package can also send commands to handle some aspects of the application:

    
    var connFu = require('connfu');
    
    // Init application
    var app = connFu.createApp('<your_api_key>');

    // Create a voice channel with a UK number
    app.commands.voice().create("conference", "uk", {welcome_message: "Welcome!"}, function(err, data){
      // Add a number to the whitelist
      app.commands.voice("conference").whitelist().create("+44xxxxxx", "rafeca", function(err, data){    
        // Update the name of the recently whitelisted number
        app.commands.voice("conference").whitelist("+4444xxxxxx").update("the wizard");
      });
    });

    /* (...) */
    
    // Delete the voice channel
    app.commands.voice("conference").delete();

For more thorough examples, look at the following examples

## Examples

[fancy-connfu-application.js](http://bluevialabs.github.com/connfu-node/fancy-connfu-application.html) - Very simple application that prints the events received from the streaming API

[growl-notifier.coffee](http://bluevialabs.github.com/connfu-node/growl-notifier.html) - CoffeeScript application that notfies via Growl the received events

[commands.js](http://bluevialabs.github.com/connfu-node/commands.html) - Express.js application that exposes all the DSL commands to a RESTful interface

## Running Tests

To run the test suite first invoke the following command within the repo, installing the development dependencies:

    
    $ npm install --dev

then run the tests:

    
    $ npm test

## License

(The MIT License)

Copyright (c) 2011 Rafael Oleza &lt;roa@tid.es&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
## Change Log

### 0.2.0 — *October 6, 2011*

  * Added `connfu` CLI to allow fast and easy application manipulation 

### 0.1.1 — *October 5, 2011*

  * `connfu-node` is now public
  * Improved documentation
  * Added `gh-pages` with a public web page that is generated automatically
  * Integration with Travis CI service

### 0.1.0 — *Sept 9, 2011*

  * Added Twitter provisioning commands
  * Added multiple phone number provisioning commands
  * Added the new optional parameters to voice channels (welcome messages and privacy)
  * Fixed bug in getAllVoiceChannels() command
  * Updated connFu hostnames to the pre-release ones
  * Improved the example applications
  * Added a new CoffeeScript example application that shows events in Growl
  * Added inline documentation to example applications
  
### 0.0.4 — *August 24, 2011*

  * Added the ability to send RSS commands via the connFu REST API
  * Standarized the format of the events data
  * Added autotest feature to ease the BDD development

### 0.0.3 — *July 28, 2011*

  * Added the ability to send voice commands via the connFu REST API

### 0.0.2 — *July 27, 2011*

  * Added a logging system to ease the debugging
  
### 0.0.1 — *July 26, 2011*

  * Initial release


spec/spies/https_get.js
var https = require('https');

var defaultHostname = 'stream.connfu.com';

exports.streamEvents = function(hostname, apiKey, events, timeBetweenEvents, endRequest, socketError) {
  
  timeBetweenEvents = timeBetweenEvents || 0;
  endRequest = endRequest || 0;
  hostname = hostname || defaultHostname;
    
  spyOn(https, 'get').andCallFake(function(options, callback){
    
    expect(options.host).toEqual(hostname);
    expect(options.headers.Authorization).toEqual('Backchat ' + apiKey);
    
    var ended = false;
    
    callback({
      statusCode: 200,
      on: function(name, finalCallback) {
        if (name == 'data') {
          // Send the data in intervals separated by time
          events.forEach(function(ev,k){
            setTimeout(function(){
              if (!ended) {
                finalCallback(new Buffer(ev));
              }
            }, (k+1)*timeBetweenEvents);
          });
        } else if (name == 'end') {
          // Only close the connection if endRequest > 0
          if (endRequest > 0 && !socketError) {
            setTimeout(function() {
              ended = true;
              finalCallback();
            }, endRequest);
          }
        }
      }
    });
    return {
      on: function(name, callback) {
        if (name == 'error') {
          if (endRequest > 0 && socketError) {
            setTimeout(function(){
              ended = true;
              callback("Socket error!");
            }, endRequest);
          }
        }
      },
      destroy: function() {
        ended = true;
      }
    };
  });
};



spec/constructor_spec.js
var connFu = require('../index');

describe('connFu basic initialization', function(){

  var app;

  beforeEach(function() {
    app = connFu.createApp("1234");
  });

  it("should initialize correctly", function(){
    expect(app).toBeDefined();
  });

  it("should initialize properly the token attribute", function(){
    expect(app.token).toEqual("1234");
  });

  it("should not be able to write the token attribute", function(){
    expect(function(){
      app.token = "1111";
    }).toThrow("Cannot set property token of #<Object> which has only a getter");
  });
  
  it("should accept a hash as a single constructor argument with all the handlers", function(){
    app = connFu.createApp({
      apiKey: "1234",
      handlers: {
        twitter: {
          new: function(){}
        },
        voice: {
          join: function(){},
          leave: function(){}
        },
        error: function(){}
      }
    });
    
    expect(app.token).toEqual("1234");
    expect(app.handlers.twitter.new).toBeDefined();
    expect(app.handlers.voice.join).toBeDefined();
    expect(app.handlers.error).toBeDefined();
  });
});

autotest.js
/**
 * Autotester to execute the tests when a file is modified and notify the results to Growl
 **/
var watch = require('nodewatch');
var growl = require('growl');
var Path= require('path');
var jasmine = require('jasmine-node');
var sys = require('sys');

growl.binVersion(function(err, version){
  if (err) {
    sys.puts('You have to install growlnotifier to be able to use autotest');
    process.exit(1);
  }

  watch.add('./spec').add('./spec/spies').add('./lib').onChange(function(file,prev,curr){

    sys.puts('File ' + file + ' modified... Executing tests...');

    var specFolder = Path.join(process.cwd(), 'spec');
    var libFolder = Path.join(process.cwd(), 'lib');
    var indexFile = Path.join(process.cwd(), 'index.js');

    // Configure jasmine-node
    jasmine.dev_mode = 'src';
    for (var key in jasmine) {
      global[key] = jasmine[key];
    }
    jasmine.loadHelpersInFolder(specFolder, new RegExp("[-_]helper\\.(js)$"));

    jasmine.executeSpecsInFolder(
      specFolder,
      function(runner, log){
        if (runner.results().failedCount == 0) {
          growl.notify(
            'All tests passed! (' + runner.results().passedCount + '/' + runner.results().totalCount + ')',
            {title: 'connFu-dsl Unit Tests', image: 'resources/tests-pass.png'}
          );
        } else {
          // Get the desc of the first failing test
          var description = '1';
          runner.results().items_.forEach(function(value){
            if (value.failedCount !== 0) {
              value.items_.forEach(function(val){
                if (val.failedCount !== 0) {
                  description = val.description;
                }
              });
            }
          });

          growl.notify(
            runner.results().failedCount + ' tests failed! (' + runner.results().passedCount + '/' + runner.results().totalCount + ')',
            {title: description, image: 'resources/tests-fail.png'}
          );
        }

        // Restart jasmine environment after executing the test suite
        var jasmineEnv = jasmine.getEnv();
        jasmineEnv.currentRunner_ = new jasmine.Runner(jasmineEnv);

        // Remove loaded files from cache
        for (var required in require.cache) {
          if (required.indexOf(specFolder) === 0 || required.indexOf(libFolder) === 0 || required.indexOf(indexFile) === 0) {
            delete require.cache[required];
          }
        }
      },
      true, // Verbose mode
      true , // Show colors
      new RegExp(".spec\\.(js)$", 'i')
    );
  });
});


lib/main.js
/**
 * Module dependencies.
 */
var connFu = require('./connFu');
var winston = require('winston');

/**
 * connFu version.
 */
exports.version = '0.2.0';

/**
 * Set up default logger
 **/
var logger = new(winston.Logger)({
  transports: [
  new(winston.transports.Console)({
    level: "info"
  })]
}).cli();

/**
 * Function to set another logger
 **/
exports.setLogger = function(newLogger) {
  logger = newLogger;
};

/**
 * Creates a new connFu application
 *
 * @param params string|array
 * @return {connFu Application}
 */
exports.createApp = function(params) {

  if (typeof params === 'string') {
    return new connFu(params, logger);
  }

  var app = new connFu(params.apiKey, logger);

  for (stream in params.handlers) {
    app.on(stream, params.handlers[stream]);
  }
  return app;
};

/**
 * Sets the global options for connFu
 *
 * @param options
 */
exports.setOptions = function(options) {
  connFu.prototype.options = options;
};


lib/connFu.js
var Client = require('./client');

/**
 * connFu application code. It is exposed directly to outside
 *
 * @param apiKey
 * @param options
 * @return {connFu Application}
 */
exports = module.exports = connFu = function(apiKey, logger) {

  // Available streams
  var _availableStreams = ['voice', 'twitter', 'error', 'sms'];

  // Default options
  var _options = {
    streamHostname: 'stream.connfu.com',
    apiHostname: 'api.connfu.com',
    apiVersion: 'v1',
    useSsl: true
  };

  // Overwrite default options with global options
  if (this.options) {
    for (var x in this.options) {
      _options[x] = this.options[x];
    }
  }

  // REST API client
  var _client;

  // Stream request object
  var _streamRequest;

  // Array to store the event handlers
  var _handlers = [];

  // is the app ending?
  var _isEnding = false;
  
  // empty function
  var noop = function(){};

  // Starts listening to a stream
  var _listenToStream = function(streamName) {
    logger.info('Connecting to connFu stream ' + streamName + '...');

    // Get port based on hostname
    var host = _options.streamHostname.split(':', 2);
    var defaultPort = _options.useSsl ? 443 : 80;

    options = {
      host: host[0],
      port: host[1] ? parseInt(host[1], 10) : defaultPort,
      path: '/' + streamName,
      method: 'GET',
      headers: {
        Authorization: 'Backchat ' + apiKey
      }
    };

    var https = _options.useSsl ? require('https') : require('http');
    _streamRequest = https.get(options, function(response) {
      // Throw an error if the HTTP status code is not valid
      if (response.statusCode > 299) {
        var msg = 'Got an HTTP error while trying to connect to the stream: HTTP Code ' + response.statusCode;
        logger.error(msg);
        (methods.handlers.error||noop)({
          message: msg,
          code: response.statusCode
        });
        return;
      }

      logger.info('Connected to connFu stream, Waiting for events...');

      response.on('data', _processStreamChunk);

      response.on('end', function() {
        // Stream dropped us... Reconnect to stream!
        if (!_isEnding) {
          logger.warn('Got disconnected from connFu stream. Reconnecting...');
          methods.listen();
        }
      });
    });

    _streamRequest.on('error', function(e) {
      // if the application is not finishing, send error and reconnect to stream!
      if (!_isEnding) {
        logger.debug('Got disconnected from connFu. Reconnecting...');
        (methods.handlers.error||noop)({
          message: 'Got an error from the stream: ' + e,
          code: 500
        });
        methods.listen();
      }
    });
  };
  
  // Processes and parses an event received from the HTTP Stream
  var _processStreamChunk = function(chunk) {
    // Trim the data received from the server and check it's not empty
    chunk = chunk.toString().match(/^\s*(.*)\s*$/)[1];
    if (!chunk) {
      return;
    }
    logger.info('Event received from connFu: ' + chunk.toString());

    // Parse the data received from connFu server
    var data = JSON.parse(chunk);

    // Hack to split between Voice events and Twitter events
    if (data instanceof Array) {
      // Voice or SMS event
      var eventName = data[0];
      var params = data[1];
      
      var channelName = null;
      var content = null;
      
      switch(eventName) {
        case 'sms':
          var channelType = 'sms';
          eventName = 'new';
          content = params.message;
          break;
        default:
          // Remove the connfu:// & /#connference stuff from the conferenceId
          var channelType = 'voice';
          channelName = params.conferenceId.match(/^[A-Z]*:\/\/([^\/#]+)/i)[1];
          break;
      }
      
      // Create the event object
      var output = {
        id: null,
        content: content,
        from: params.from,
        to: params.to,
        messageType: eventName,
        channelType: channelType,
        channelName: channelName,
        timeStamp: null
      };
      if (methods.handlers[channelType]) {
        (methods.handlers[channelType][eventName]||noop)(output);
      } 
    } else {
      // Twitter event
      
      // Create the event object
      var output = {
        id: data.object.id,
        content: data.object.content,
        from: sender = data.actor.id,
        to: data.object.entities.user_mentions ? data.object.entities.user_mentions : [],
        messageType: "new",
        channelType: "twitter",
        timeStamp: new Date(Date.parse(data.object.published))
      };
      if (methods.handlers.twitter) {
        (methods.handlers.twitter['new']||noop)(output);
      }
    }
  };

  /**
   * Public interface
   **/
  var methods = {
    // Method used to add event handlers
    on: function(stream, events) {
      // Check if the stream name is valid
      if (_availableStreams.indexOf(stream) === -1) {
        throw('Stream name invalid');
      }

      // Initialize stream handler
      _handlers[stream] = events;

      // Implement fluent interface
      return methods;
    },

    // Listen method
    listen: function() {
      logger.debug('Starting listen()...');

      // Get the stream name using the REST API
      methods.commands.getApp(
        function(err, app) {
          if (err) {
            logger.error('Error getting the stream name', err);
            (methods.handlers.error||noop)({
              message: 'Got a HTTP ' + err.code + ' error while trying to get the stream name',
              code: err.code
            });
          } else {
            _listenToStream('connfu-stream-' + app.name);            
          }
        }
      );

      // Implement fluent interface
      return methods;
    },

    // End connection with connfu streaming server
    end: function() {
      logger.debug('Disconnecting from connFu...');
      _isEnding = true;
      _streamRequest.destroy();
    },

    // Accessor to apiKey
    get token() {
      return apiKey;
    },

    // Accessor to REST API Client
    get commands() {
      if (!_client) {
        return new Client(apiKey, logger, _options);
      }
    },

    // Accessor to defined handlers
    get handlers() {
      return _handlers;
    }
  };

  return methods;
};


examples/fancy-connfu-application.js
// Don't be mistaken by the name, this is a very stupid example of how should a 
// **connFu** application look like. This application only listens to voice
// and twitter events and prints them on the terminal.
//
// Running this example
// --------------------
//
// * First of all, you have to register into [connFu.com](https://connfu.com)
// an application. You will receive an **application token** which will be very useful ;)
//
// * Then, install the connFu npm package:
//
//          $ npm install connfu
//
// * Then get the source of this example from [here](https://github.com/bluevialabs/connfu-node/blob/master/examples/fancy-connfu-application.js)
//   and run it!
//
//          $ node fancy-connfu-application.js <application_token>

// Application code explanation
// ----------------------------

// As we plan to launch this application via the command line,
// we should receive the application token as an argument, so we need 
// to check that the application token is passed as a parameter
if (process.argv.length < 3) {
  console.log("Usage: node fancy-connfu-application.js <application_token>");
  process.exit(1);
}

//### Initializing the connFu application

// First of all, we have to require the connFu package...
//
// ...and then initialize the connFu application using the application
//token received from the command line
var connFu = require('connfu');

var app = connFu.createApp(process.argv[2]);

//### Adding a voice channel listener
//
// Voice channels have three different events: `join`, `leave` and `change_topic`.
// Inside the event listener we can add different handlers for each of those events
//
// The `params` received in the event handler is an array with all the information about the event.
// Check the docs for more detailed info about this object

// Add the voice channel listener
app.on('voice', {
  // Add `join` event handler
  join: function(params) {
    console.log('The number ' + params.from + ' joined to the conference with number ' + params.to);
  },

  // And then a `leave` event handler
  leave: function(params) {
    console.log('The number ' + params.from + ' left the conference with number ' + params.to);
  }
});

//### Adding a Twitter channel listener

// We can add a Twitter event listener very easily too:
app.on('twitter', {
  
  // Twitter channels only have the `new` event, which is received when a tweet is received
  new: function(data) {
    console.log(data);
  }
});

//### Start listening to the channels

// This is how we start listening! Pretty easy, huh?
app.listen();


lib/client.js
var querystring = require('querystring');

/**
 * connFu application code. It is exposed directly to outside
 *
 * @param apiKey
 * @param options
 * @return {connFu api client}
 */
exports = module.exports = client = function(apiKey, logger, options) {

  var _logger = options.logger;

  var _request = function(method, path, data, callback) {
    // Define default callback
    callback = callback || function() {};

    // Add data to the query string (GET & DELETE requests)
    if (data.length && ['GET', 'DELETE'].indexOf(method) != -1) {
      path += '?' + querystring.stringify(data);
    }

    var headers = {
      AUTH_TOKEN: apiKey,
      'Content-Type': 'application/json',
      Accept: 'application/json'
    };
    // Add conent-length header for POST & PUT requests
    if (['POST', 'PUT'].indexOf(method) != -1) {
      headers['Content-Length'] = JSON.stringify(data).length;
    } else {
      headers['Content-Length'] = 0;
    }

    // Get port based on hostname
    var host = options.apiHostname.split(':', 2);
    var defaultPort = options.useSsl ? 443 : 80;

    var params = {
      host: host[0],
      port: host[1] ? parseInt(host[1], 10) : defaultPort,
      method: method,
      path: '/' + options.apiVersion + path,
      headers: headers
    };

    logger.info('Making HTTP request: ' + method + ' http://' + params.host + ':' + params.port + params.path);

    var https = options.useSsl ? require('https') : require('http');
    var req = https.request(params, function(res) {
      var output = '';
      var sys = require('sys');

      // Read all the data from the response
      res.on('data', function(chunk) {
        output += chunk;
      });

      // Send the response to the callback
      res.on('end', function() {
        // If the status code is different than 2xx throw an error
        if (res.statusCode > 299) {
          logger.error('Server returned a ' + res.statusCode + ' HTTP error. Body: ' + output);
          callback({
            message: 'Server returned a ' + res.statusCode + ' HTTP error',
            code: res.statusCode
          });
          return;
        }

        logger.info('Response from server: ' + output);

        // Trim the data received from the server
        output = output.toString().match(/^\s*(.*)\s*$/)[1];

        // Parse & return in the callback the JSON response
        if (output) {
          callback(null, JSON.parse(output));
        } else {
          // When creating resources, return the "location" header
          if (res.statusCode === 201) {
            callback(null, res.headers.location);
            return;
          }
          callback();
        }
      });

      res.on('error', function(e) {
        logger.error('There was an error making the request: ' + e);
        callback({
          message: e.message,
          code: e.code
        });
      });
    });

    req.on('error', function(e) {
      logger.error('There was an error making the request: ' + e);
      callback({
        message: e.message,
        code: e.code
      });
    });

    // Add data to the body of the request (POST & PUT requests)
    if (data && ['POST', 'PUT'].indexOf(method) != -1) {
      req.write(JSON.stringify(data));
    }

    req.end();
  };

  var _get = function(path, data, callback) {
    _request('GET', path, data, callback);
  };

  var _put = function(path, data, callback) {
    _request('PUT', path, data, callback);
  };

  var _post = function(path, data, callback) {
    _request('POST', path, data, callback);
  };

  var _delete = function(path, data, callback) {
    _request('DELETE', path, data, callback);
  };

  return {
    /**
     * Returns the app information
     **/
    getApp: function(callback) {
      _get('/', {}, callback);
    },

    voice: function(voiceName) {
      return {
        // client.voice(name).get()
        get: function(callback) {
          var path = '/channels/voice';
          path += voiceName ? '/' + voiceName : '';
          _get(path, {}, callback);
        },

        // client.voice(name).update(params)
        update: function(params, callback) {
          _put('/channels/voice/' + voiceName, params || {}, callback);
        },

        // client.voice().create(name, country, params)
        create: function(voiceName, country, params, callback) {
          if (!callback) {
            callback = params;
            params = {};
          }
          params.uid = voiceName;
          params.country = country;
          _post('/channels/voice', params, callback);
        },

        // client.voice(name).remove()
        remove: function(callback) {
          if (!voiceName) {
            throw("channel name must be provided");
          }
          _delete('/channels/voice/' + voiceName, {}, callback);
        },

        whitelist: function(phone) {
          var path = '/channels/voice/' + voiceName + '/whitelisted';
          path += phone ? '/' + querystring.escape(phone) : '';
          
          return {
            // client.voice(name).whitelist().get()
            get: function(callback) {
              _get(path, {}, callback);
            },
            
            // client.voice(name).whitelist(phone).update(name)
            update: function(name, callback) {
              _put(path, {
                name: name
              }, callback);
            },

            // client.voice(name).whitelist().create(phone, name)
            create: function(phone, name, callback) {
              _post(path, {
                name: name,
                phone: phone + ''
              }, callback);
            },

            // client.voice(name).whitelist(phone).remove()
            remove: function(callback) {
              _delete(path, {}, callback);
            }
          };
        },
        
        phones: function(phoneNumber) {
          
          return {
            // client.voice(name).phones().get()
            get: function(callback) {
              _get('/channels/voice/' + voiceName + '/phones', {}, callback);
            },
            
            // client.voice(name).phones().create(country)
            create: function(country, callback) {
             _post('/channels/voice/' + voiceName + '/phones', {
                country: country
              }, callback);
            },
            
            // client.voice(name).phones(phoneNumber).remove()
            remove: function(callback) {
              _delete('/channels/voice/' + voiceName + '/phones/' + querystring.escape(phoneNumber), {}, callback);
            }
          }
        }
      };
    },
    
    rss: function(rssName) {
      return {
        // client.rss(name).get()
        get: function(callback) {
          var path = '/channels/rss';
          path += rssName ? '/' + rssName : '';
          _get(path, {}, callback);
        },

        // client.rss(name).update(newUri)
        update: function(newUri, callback) {
          _put('/channels/rss/' + rssName, {
            uri: newUri
          }, callback);
        },

        // client.rss().create(uri)
        create: function(rssName, uri, callback) {
          _post('/channels/rss', {
            uid: rssName,
            uri: uri
          }, callback);
        },

        // client.rss(name).remove()
        remove: function(callback) {
          _delete('/channels/rss/' + rssName, {}, callback);
        }
      };
    },
    
    twitter: function(twitterName) {
      return {
        // client.twitter(name).get()
        get: function(callback) {
          var path = '/channels/twitter';
          path += twitterName ? '/' + twitterName : '';
          _get(path, {}, callback);
        },

        // client.twitter().create(uri, "origin",  ["rafeca"], filters)
        create: function(uid, type, accounts, filters, callback) {
          var params = {
            uid: uid
          };
          
          var appliedFilters = [];
          
          if (type === 'origin') {
            params.accounts = accounts.map(function(account){
              return {name: account};
            });
          } else if (type === 'mentioned') {
            if (typeof(accounts) !== 'string') {
              throw('Only one mentioned account is allowed to be added in a Twitter channel');
            }
            params.accounts = [{name: accounts}];
            appliedFilters.push('recipients:' + accounts);
          } else {
            throw("Invalid account type. Valid types are: 'origin', 'mentioned'");
          }
          
          if (filters) {
            for (var k in filters) {
              appliedFilters.push(k + ':(' + filters[k].join(' AND ') + ')');
            }
          }
          
          if (appliedFilters.length) {
            params.filter = appliedFilters.join(' AND ');
          }
          
          _post('/channels/twitter', params, callback);
        },

        // client.twitter(name).remove()
        remove: function(callback) {
          _delete('/channels/twitter/' + twitterName, {}, callback);
        }
      };
    }
  };
};
